### PR TITLE
29 document permissions

### LINKED ISSUE #29: Document permissions
There are two permissions required to access a patient document:
- [ ] Patient access: user must have access to patient the document belong too
- [x] Document type access: user must have QUERY and/or MUTATE access for the document type

## Patient access:
In a first step we assume the user only has access to all patients at stores the user has access too.
The name_store_join table can be consolidate to find out if a patient is available at a store.

How to find out which document belongs to which patient?
There are three options:
- ~~introduce a patient_document_join table~~
- ~~just by document name, e.g. by convention patient document names already starts with the patient id~~
- store patient id on document table as an optional column (this comes at the cost that the document table becomes somewhat patient specific) Or better call it document `owner` which is more generic.

- [x] #590 
## Document type access
A new table is needed to store document type permissions.

- [x] #593 

# Protecting Endpoints
- [x] Ensure that the document_registry endpoints only returns document types the user has access to
- [x] Check permissions on mutation/query endpoints

### PR DESCRIPTION
**Pull request by roxy-dao**
Closes #29 

Note: Might not compile and will need changes from #609 

- [ ] Query and mutation endpoints for documents check the user's context before allowing them to mutate/query endpoints
- [ ] Added r#type to UpdateEncounter... but can remove if this shouldn't be here...
- [ ] Does the insert Json form schema endpoint also need to be locked behind context?

### CODE REVIEW

**Review by mark-prins — COMMENTED** [review]
Unable to test.. as it will need the table changes merged first.
Overall, looks good! just wondering about the filter and if specifying a filter in a query will return documents which the user won't have access to?

File: server/graphql/programs/src/queries/program_enrolment.rs [hunk 1]
```diff
@@ -77,13 +79,20 @@ pub fn program_enrolments(
     let service_provider = ctx.service_provider();
     let context = service_provider.basic_context()?;
 
+    let filter =
```
[the remark this answers is not in the document]
Comment (roxy-dao) [reply server/graphql/programs/src/queries/program_enrolment.rs]: [d0d0680](https://github.com/openmsupply/open-msupply/pull/611/commits/d0d06805ff981b21b0c11e780e0f8d8721cd9be6)

**Review by mark-prins — COMMENTED** [review]
For the site to work I've had to add a permission row for `DOCUMENT_PROGRAM_QUERY` for every type on the document table. I'm hoping that will come through in the sync..? I'm wondering if a single string for the context is just too limited.

More importantly, the programs query fails with these permission changes - so we cannot enrol patients in programs!

<details>
<summary>Query</summary>

```
query programs($key: DocumentRegistrySortFieldInput!, $desc: Boolean) {
  documentRegistries(
    sort: {key: $key, desc: $desc}
    filter: {context: {equalTo: PROGRAM}}
  ) {
    ... on DocumentRegistryConnector {
      __typename
      totalCount
      nodes {
        __typename
        ...ProgramDocumentRegistry
      }
    }
  }
}

fragment ProgramDocumentRegistry on DocumentRegistryNode {
  ...ProgramDocument
  children {
    ...ProgramDocument
  }
}

fragment ProgramDocument on DocumentRegistryNode {
  id
  documentType
  formSchemaId
  jsonSchema
  name
  context
  parentId
  uiSchema
  uiSchemaType
}
```
</details>

<details>
<summary>Variables</summary>

```
{
  "key": "documentType"
}
```
</details>

Returns this along with the `Forbidden` error:

```
"extensions": {
        "details": "Store id not specified in request, Required permissions: And([HasStoreAccess, HasPermission(DocumentProgramQuery)])"
      }
    }
  ]
```

**Comment by roxy-dao** [comment]
> For the site to work I've had to add a permission row for `DOCUMENT_PROGRAM_QUERY` for every type on the document table. I'm hoping that will come through in the sync..? I'm wondering if a single string for the context is just too limited.
> 
> More importantly, the programs query fails with these permission changes - so we cannot enrol patients in programs!
> 
> Query
> ```
> query programs($key: DocumentRegistrySortFieldInput!, $desc: Boolean) {
>   documentRegistries(
>     sort: {key: $key, desc: $desc}
>     filter: {context: {equalTo: PROGRAM}}
>   ) {
>     ... on DocumentRegistryConnector {
>       __typename
>       totalCount
>       nodes {
>         __typename
>         ...ProgramDocumentRegistry
>       }
>     }
>   }
> }
> 
> fragment ProgramDocumentRegistry on DocumentRegistryNode {
>   ...ProgramDocument
>   children {
>     ...ProgramDocument
>   }
> }
> 
> fragment ProgramDocument on DocumentRegistryNode {
>   id
>   documentType
>   formSchemaId
>   jsonSchema
>   name
>   context
>   parentId
>   uiSchema
>   uiSchemaType
> }
> ```
> 
> Variables
> ```
> {
>   "key": "documentType"
> }
> ```
> 
> Returns this along with the `Forbidden` error:
> 
> ```
> "extensions": {
>         "details": "Store id not specified in request, Required permissions: And([HasStoreAccess, HasPermission(DocumentProgramQuery)])"
>       }
>     }
>   ]
> ```
It should come from sync once the tables and permissions have been added....

About the single string for context comment... maybe https://github.com/openmsupply/open-msupply/issues/596 this can solve it? 🤔 

The permissions probably didn't get saved into the DB since it re-syncs and new pull doesn't have the permissions... 
I'll put them into init_programs_data for testing?

**Comment by mark-prins** [comment]
aha, I see now! thanks yes - #596 sounds like it would address it 😁 

And yes - if you could add the permissions to the init_programs_data - that would be helpful. It makes the set up easier if the site works out of the box like that

**Comment by roxy-dao** [comment]
> aha, I see now! thanks yes - #596 sounds like it would address it 😁
> 
> And yes - if you could add the permissions to the init_programs_data - that would be helpful. It makes the set up easier if the site works out of the box like that

Found it after we talked... please disregard that conversation 😭 
[ae44031](https://github.com/openmsupply/open-msupply/pull/611/commits/ae44031dff3e41f4a76ea45fdaeb1656f7edaf48)

File: server/service/src/sync/init_programs_data.rs [hunk 2]
```diff
@@ -457,6 +458,48 @@ fn encounter_hiv_care_5() -> hiv_care_encounter::HivcareEncounter {
     })
 }
 
+pub fn insert_programs_permissions(connection: &StorageConnection, user_id: String) {
+    UserPermissionRowRepository::new(&connection)
+        .upsert_one(&UserPermissionRow {
+            id: uuid(),
+            user_id: user_id.clone(),
+            store_id: Some("D77F67339BF8400886D009178F4962E1".to_string()),
```
Comment (mark-prins) [inline server/service/src/sync/init_programs_data.rs:467]: the hard coded store id is an issue for me. With my configuration I cannot actually switch to this store, so I cannot access anything 😢 

I used a postgres script to give myself permission for every store which the user has access to ( from `user_store_join` ).

File: server/service/src/sync/init_programs_data.rs [hunk 3]
```diff
@@ -457,6 +458,48 @@ fn encounter_hiv_care_5() -> hiv_care_encounter::HivcareEncounter {
...
+        .upsert_one(&UserPermissionRow {
+            id: uuid(),
+            user_id,
+            store_id: Some("D77F67339BF8400886D009178F4962E1".to_string()),
+            permission: Permission::DocumentProgramMutate,
+            context: Some("Patient".to_string()),
+        })
+        .unwrap();
+}
+
```
Comment (mark-prins) [inline server/service/src/sync/init_programs_data.rs:503]: I'm being asked for `DocumentEncounterQuery` permission too - this is to view the list of encounters. Do you get that as well? or is the above enough?

**Review by mark-prins — COMMENTED** [review]
Changes look good thanks! I was looking to do the same kind of thing 😁 

Just concerned that it should add all stores, not just the Tamaki store id

File: server/service/src/sync/init_programs_data.rs (hunk 3 above)
Comment (roxy-dao) [reply server/service/src/sync/init_programs_data.rs]: [b5f8c8b](https://github.com/openmsupply/open-msupply/pull/611/commits/b5f8c8b3f42060c70799ff2fcdad6d3f2107ed42) 

Inserted permissions for all stores since couldn't add store id from the `update_user` 🤔 (or maybe I might have missed something).... more things to be deleted later.... 😰

File: server/graphql/programs/src/mutations/encounter/insert.rs [hunk 4]
```diff
@@ -45,6 +45,14 @@ pub fn insert_encounter(
     let service_provider = ctx.service_provider();
     let service_context = service_provider.basic_context()?;
 
+    match user.context.into_iter().find(|c| c == &input.r#type) {
```
Comment (andreievg) [inline server/graphql/programs/src/mutations/encounter/insert.rs:48]: this sort of mapping is usually done with [ok_or](https://doc.rust-lang.org/std/option/enum.Option.html#method.ok_or), we use it in service validations and in some places in graphql

Comment (andreievg) [reply server/graphql/programs/src/mutations/encounter/insert.rs:48]: Hmm, so the context in this case would be any context (both mutate and query) from all of the user permissions, there are a number of ways to solve this:
* Pass context through to validation through a string enum variant: `Resource::MutateEncounter(String)`, and make `HasPermissionWithContext { permisison, context }`, but existing DSL would need to be refactored quite a bit to pass `context` from MutateEncounter `Resource` variant to HasPermissionWithContext variant of `PermissionDSL` : /
* Another way is for `validate_resource_permissions` to return array of context that match permission type, but because validate_resource_permissions is nested with possible `AND` or `OR` being the parent, it seems tricky and hacky and would only work for one uses case where there is no OR and AND in permission DSL variants.

I would make and issue with permission improvements shortly, it would question the need for permission DSL and how much complexity are we willing to accept by continuing to squeeze functionality into it. I would prefer for this to be a very simple validation done in `validate` method in service (with some helpers)

Like this: https://github.com/openmsupply/open-msupply/compare/29-Document-permissions...29-Validation-in-service

**Review by andreievg — COMMENTED** [review]
We shouldn't have permissions for any update/inserts of document registry/json schema, so I think the following should be either some `Dev` permission or be commented to be removed:

```
    MutateDocumentRegistry,
    MutateJsonSchema,
```

Just fyi it sounds like we would just have ProgramType permissions, rather then the granular permissions we have, ProgramType permission will still be have dynamic context btw.

File: server/graphql/programs/src/mutations/encounter/insert.rs (hunk 4 above)
Comment (jmbrunskill) [reply server/graphql/programs/src/mutations/encounter/insert.rs:48]: The example with validation in the service layer looks nice to me.
I think it could still be combined with DSL If needed for a particular use case. No reason (that I'm aware of) that the DSL is limited to Graphql layer.

Comment (clemens-msupply) [reply server/graphql/programs/src/mutations/encounter/insert.rs:48]: I read this too late and already toyed around with option 2. The `AND` / `OR` cases where actually not too bad to handle and I made `validate_resource_permissions` return list of permissions needed for this endpoint. However, its still a bit tricky to use and after discussion with @andreievg we thought it might be easier to have a helper method that is more straight forward. 

Regarding, service validation. I like the separation of concern between user permissions and services, i.e. the permission system is independent from the service logic. This makes it possible to use the service in other places without running into permission issues, e.g. a cleanup, sync or migration task can use the service. Moreover, we could make changes to the user permissions without affecting the services. Its also not too nice to have to read the user permissions in two different places, i.e. in the DSL and in a service.

However, I think we can find a good compromise here. I would keep reading and handling the user permission in the graphql layer next to the DSL validation and extract a list of relevant contexts using a helper method. This list is then passed on to the service and the service can do it's validation. For example,
```rust
// in graphql:
let validated_user = validate(...)?;
let allowed_doc_types: Vec<String> = helper_get_permissions_contexts(Permission::DocumentEncounterQuery, validated_user.user_permissions);

service.query(allowed_doc_types, ...)
```

In other words we extract the user's capabilities from the user permissions and pass them on to the service and the service can do its validation based on these capabilities.

**Comment by clemens-msupply** [comment]
> Unable to test.. as it will need the table changes merged first. Overall, looks good! just wondering about the filter and if specifying a filter in a query will return documents which the user won't have access to?

I actually have a string filter helper that restricts the filter results to an allowed list of strings. I also have it reasonably well tested. I plan to add it soon.

**Comment by clemens-msupply** [comment]
Looks good to me and thing we can merge and address identified issues in a new PR:

- simplify permissions -> reduce number of program related permissions
- do some validation in services

**Review by clemens-msupply — APPROVED** [review]
